### PR TITLE
[REG-1293] Update Agent Builder documentation for Unity SDK

### PR DESCRIPTION
**Pull request by abeizer**
* Moved some pages into a /common directory to share between `players` and `studios` routes
* Added "Agent Builder" page to `studios` sidebar
* Updated images of Agent Builder's UI, since it has changed a bit since the last time we updated the docs
  * Added new images for Unity bots
* Reorganized `studios` sidebar a bit to make it flow more nicely
  * Updated alllllll the links... this was killer
Screenshot 2023-10-20 at 1 51 57 PM

### CODE REVIEW

**Comment by RG-nAmKcAz** [comment]
Is there a loom with the new content ?

Why did we drop the top level `Unity` header in the sidebar ?  I'm sure we will need to support other engines going forward.

File: docs/common/agent-builder/_adding-nodes.mdx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Whenever you create a fresh Agent Builder project for your AI, we'll start you out with a root node and a sequence node to form the basis of your tree.
```
Comment (addisonbgross) [inline docs/common/agent-builder/_adding-nodes.mdx:1]: I think this can be a bit more concise

```suggestion
New Agent Builder projects start with a root node and a sequence node to form the basis of your tree.
```

File: docs/common/agent-builder/_behavior-trees.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+- **Failure** -> The node has failed its task. This doesn't necessarily mean an error occurred. For example, a heal action might fail if the character doesn't have any potions.
+
+Different types of nodes have different responsibilities and react differently to the states of other nodes.
+
+- **Root Node**: The beginning of the tree, where the AI starts its decision-making.
+  Its job is simply to execute whatever node is below it in the tree and return its state.
+- **Sequence Node**: This node has one or more children and acts as a logical AND.
+  It executes its child nodes one at a time, and if any of them fail then the Sequence Node immediately fails.
+- **Selector Node**: This node also has one or more children, but acts as a logical OR.
+  It executes each of its children and if one of them succeeds, then the Selector Node immediately succeeds.
```
Comment (addisonbgross) [inline docs/common/agent-builder/_behavior-trees.md:19]: Can we add more clarity around what `success` and `failure` mean in this context? Maybe a little example

File: docs/common/agent-builder/_saving-your-tree.mdx [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Save your tree early and often! Any changes made to your AI will be lost if you leave the Agent Builder without saving.
```
Comment (addisonbgross) [inline docs/common/agent-builder/_saving-your-tree.mdx:1]: This is irrelevant to this PR, but I think it would be nice to use the History API to add a little "Are you sure?" modal before navigating back, or closing the tab, with your browser.

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+const myBool = this.getData<boolean>("myBool");
+```
+
+Whenever you submit a prompt to GPT, we supplement it with context about your game's available actions and state attributes as well as our bot API.
+However, large language models like GPT are still in their infancy and may output code that doesn't follow our expected format, or code that functions differently in-game than your prompt intends.
+You may need to alter your prompt and re-submit it for new output when this happens, or you can fix the output manually.
+
+### Validating Code
+
+The embedded code editor offers standard syntax highlighting and Intellisense features for Regressions Games APIs.
```
Comment (addisonbgross) [inline docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx:183]: Less brand specifics may be nice

```suggestion
The embedded code editor offers standard syntax highlighting and intelligent code completion features for Regressions Games APIs.
```

**Review by addisonbgross — COMMENTED** [review]
Just a few comments. These are some nice, descriptive docs

**Comment by RG-nAmKcAz** [comment]
I think we need to make some changes to clarify or get some consistency with the language around `Extract Data` and `Upload Generated Game Context`.

Something about the word `generated` being there doesn't jive with me.

Maybe these 2 phrasings as a starting point so their connection is more explicit.
`Extract Game Context`
`Upload Game Context`

**Review by vontell — COMMENTED** [review]
Approving in spirit, but want to give others a chance to approve rather than myself as well, so just adding this as a comment!

File: docs/common/agent-builder/_adding-nodes.mdx (hunk 1 above)
Comment (abeizer) [reply docs/common/agent-builder/_adding-nodes.mdx]: Updated

File: docs/common/agent-builder/_behavior-trees.md (hunk 2 above)
Comment (abeizer) [reply docs/common/agent-builder/_behavior-trees.md]: Added a practical example to this section

File: docs/common/agent-builder/_saving-your-tree.mdx (hunk 3 above)
Comment (abeizer) [reply docs/common/agent-builder/_saving-your-tree.mdx:1]: Agree, this is something [we've logged](https://linear.app/regression-games/issue/REG-1038/bt-gpt-agent-builder-do-not-allow-navigation-with-unsaved-changes) but haven't prioritized

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx (hunk 4 above)
Comment (abeizer) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: Updated

**Review by addisonbgross — APPROVED** [review]

**Comment by RG-nAmKcAz** [comment]
> > I think we need to make some changes to clarify or get some consistency with the language around `Extract Data` and `Upload Generated Game Context`.
> > Something about the word `generated` being there doesn't jive with me.
> > Maybe these 2 phrasings as a starting point so their connection is more explicit. `Extract Game Context` `Upload Game Context`
> 
> I'm always for simplifying our labels + user-facing concepts. We'll have to update docs after making changes to the actual projects so they're consistent. Tagging @rcornwall-reg for Unity-side

Please open a task for this as the work can be done on one task to update unity/ui/doc.  The screenshots will be the most 'work' of the task, but other 2 parts are 1 liner wording updates

**Comment by abeizer** [comment]
> Upload Game Context

Added [here](https://linear.app/regression-games/issue/REG-1364/update-wording-around-generating-game-context-for-agent-builder)

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+```
+
+The above outputs are good examples of how properly-formatted code should look when you're finished generating or writing it.
+
+You may notice that these examples invoke the `getData` and `setData` methods.
+These can be used to share data between nodes, and it's a good practice to specify a data type for `getData` to take advantage of the editor's Intellisense.
+
+:::note
+Classes to represent available state variables for entities are coming soon
+:::
```
Comment (analogrelay) [inline docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx:164]: Looking at the code, I think this is a little more accurate (I noticed that the data appears to be shared across ticks too and that was initially surprising to me, but I think it's useful):

```suggestion
These can be used to share data between nodes and across each tick of your bot.

:::note
We recommend you specify a data type in `getData` so you can take full advantage of autocomplete in the editor.
:::
```

And again, no need for a "Coming Soon" here I think. We don't know when "soon" is, and when "soon" does come, it'd be better if we don't have to update this note.

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+# Creating AIs Using the Agent Builder
+
+Regression Games' Agent Builder is an easy-to-use visual editor for constructing AI behavior trees.
+Our editor utilizes large language models like [OpenAI's GPT-4](https://openai.com/blog/chatgpt) to write code for you.
+This makes the Agent Builder an excellent tool for writing bots as you familiarize yourself with your project's SDK integration.
+
+:::note
+This tool is early in development - it is prone to change and may contain bugs.
+Please report issues and feature requests for the Agent Builder [here](https://regression-games.sleekplan.app/feedback).
```
Comment (analogrelay) [inline docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx:18]: One thing I've learned from writing docs is to avoid writing content you **know** will change over time, if you *can* avoid it. 

Also "here" links are less accessible, as screen readers don't do a great job of describing them. It's better to give the link text that is specific to where the link goes, rather than relying on previous context in the sentence or paragraph.
 
So maybe rephrase this as:

```suggestion
We're constantly working to improve Agent Builder. You can report issues and submit feature requests on our [Feedback Site](https://regression-games.sleekplan.app/feedback).
```

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+### Adding Nodes
+
+<AddingNodes />
+
+### Generating Code
+
+:::note
+All samples in this section are written in Typescript, which is the only language currently supported by the Agent Builder.
+If you'd like to see support for other languages, like C#, let us know [here](https://regression-games.sleekplan.app/feedback).
```
Comment (analogrelay) [inline docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx:83]: To my point above, we don't know if/when C# support will land. And while we certainly should update the docs when it does, it's easiest to just describe the situation now rather than predicting the future. We are, and always will be, "constantly evaluating support for other languages".

```suggestion
All samples in this section are written in TypeScript.
We're constantly evaluating support for other languages. Feel free to request others on our [Feedback Site](https://regression-games.sleekplan.app/feedback)!.
```

File: docs/studios/unity/unity-sdk/creating-bots/javascript/configuration.mdx [hunk 8]
```diff
@@ -15,7 +15,7 @@ make decisions, and interact with game elements.
 
 ## Project Setup
 
-A Bot project is structured as an npm package.
+A Javascript Bot project is structured as an npm package.
```
Comment (analogrelay) [inline docs/studios/unity/unity-sdk/creating-bots/javascript/configuration.mdx:18]:
```suggestion
A JavaScript Bot project is structured as an npm package.
```

File: docs/common/agent-builder/_behavior-trees.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+- **Running** -> The task represented by this node's branch hasn't yet completed.
+- **Success** -> The node has completed its task successfully.
+- **Failure** -> The node has failed its task. This doesn't necessarily mean an error occurred. For example, a heal action might fail if the character doesn't have any potions.
+
+Different types of nodes have different responsibilities and may react differently to the statuses of other nodes.
+
+- **Root Node**: The beginning of the tree, where the AI starts its decision-making.
+  Its job is simply to execute whatever node is below it in the tree and return its status.
+- **Sequence Node**: This node has one or more children and acts as a logical AND.
+  It executes its child nodes sequentially, and if any one of them fail then the Sequence Node skips execution of the remaining children and returns a failure status.
```
Comment (analogrelay) [inline docs/common/agent-builder/_behavior-trees.md:17]:
```suggestion
  It executes its child nodes sequentially, and if any one of them fail then the Sequence Node immediately returns the Failure status and stops executing child nodes.
```

**Review by analogrelay — APPROVED** [review]
Looks great! I had some copy-editing nitpicks, but the core content is great. My comments come largely from opinion and just how it "reads" to me, so please feel free to carry on with a different approach!

File: docs/common/agent-builder/_behavior-trees.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+- **Selector Node**: This node also has one or more children, but acts as a logical OR.
+  It executes its child nodes sequentially, and if any one of them succeed then the Selector Node returns a success status.
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/common/agent-builder/_behavior-trees.md]: Reworded for clarity, although slightly differently than your suggestion here

File: docs/players/creating-bots/agent-builder.mdx [hunk 11]
```diff
@@ -197,12 +151,10 @@ This is a good example of how GPT sometimes needs additional coaching to reach y
 
 ### Validating Code
 
-The embedded code editor offers standard syntax highlighting and Intellisense features for Regressions Games APIs. 
+The embedded code editor offers standard syntax highlighting and intelligent code completion features for Regressions Games APIs.
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/players/creating-bots/agent-builder.mdx:164]: This should've been addressed already 🤔 
I made sure there are no other references to "Intellisense" left

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+import BehaviorTree from '../../../../../common/agent-builder/_behavior-trees.md'
+import LoadingYourTree from '../../../../../common/agent-builder/_loading-your-tree.mdx'
+import AddingNodes from '../../../../../common/agent-builder/_adding-nodes.mdx'
+import SavingYourTree from '../../../../../common/agent-builder/_saving-your-tree.mdx'
+
+# Creating AIs Using the Agent Builder
+
+Regression Games' Agent Builder is an easy-to-use visual editor for constructing AI behavior trees.
+Our editor utilizes large language models like [OpenAI's GPT-4](https://openai.com/blog/chatgpt) to write code for you.
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: I like your suggestion - made some tweaks to it but the flow should be similar

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+  }
+}
+```
+
+:::info
+All of your node's code must be defined within the `LeafNode` class, and the `execute` method must be defined as your entrypoint into the node.
+You may encounter errors loading your AI into a game if these are not properly defined when you save your tree.
+:::
+
+You can either manually type code into the editor (see the [PlayTest Bots API](./playtest-bots)), or you can have GPT do the heavy lifting for you.
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: Updated all references to GPT so they're now some permutation of "AI code assistant"

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx (hunk 6 above)
Comment (abeizer) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: Updated!

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx (hunk 7 above)
Comment (abeizer) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: Updated! I believe Sleekplan has a public roadmap feature. I don't think we're currently using it, but it could be really helpful for simplifying our documentation. ex. "See our roadmap for language support" (@vontell)

File: docs/common/agent-builder/_behavior-trees.md (hunk 9 above)
Comment (abeizer) [reply docs/common/agent-builder/_behavior-trees.md]: Updated, kept the wording similar to the updated verbiage for Selector Node

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx (hunk 5 above)
Comment (abeizer) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: This is the only comment that I disagree with. I think it's going to be very confusing for someone to read "make sure you specify a type!" in the docs when the relevant types don't exist yet. We need some indication that we recommend the use of types where they _can_ be applied, but that we're still working on implementing types to represent your states and actions

Comment (abeizer) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: Maybe "under development" is a better phrase since you're right that we don't know how _soon_ "soon" is

Comment (analogrelay) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: > I think it's going to be very confusing for someone to read "make sure you specify a type!" in the docs when the relevant types don't exist yet.

A very fair point, though perhaps one that argues for just omitting the "make sure you specify a type" comment until we actually have the type infrastructure in place?

Either way, I'm OK with it. I'm certainly not significantly opposed to referencing future work. I think in a previous iteration (before I saved the comment) I had something like this:

```suggestion
These can be used to share data between nodes and across each tick of your bot.

:::note
We recommend you specify a data type in `getData` so you can take full advantage of autocomplete in the editor. The types available in the editor are limited, but we're continually adding more!
:::
```

File: docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx (hunk 7 above)
Comment (analogrelay) [reply docs/studios/unity/unity-sdk/creating-bots/javascript/agent-builder.mdx]: That's a great way to do this. Centralizing where we talk about the future makes it easier to ensure we keep it relatively up-to-date :). My main concern is having lots of different docs referencing future features that may or may not actually exist at the time the reader is looking at the page.

File: docs/common/agent-builder/_behavior-trees.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/common/agent-builder/_behavior-trees.md]: Looks great!
